Add switchable themes and Dark theme

tsonnen opened the pull request:
- Colors are a suggestion, all of them are easily editable in helpers/theme.dart
- Created a helpers folder for the theme data
- Created a settings screen for users to switch the theme easily
    - Functionality could be moved to the app drawer (or anywhere) as well/instead of if preferred.
- Added [Pref](https://pub.dev/packages/pref) to allow for saving preferences
- Added relevant test

Fixes #7

hanskokx commented on lib/helpers/theme.dart:
```diff
@@ -0,0 +1,56 @@
+    switch (theme) {
+      case ThemeEnum.lightTheme:
+        return lightTheme;
+        break;
+      case ThemeEnum.darkTheme:
+        return darkTheme;
+        break;
+      default:
+        break;
```
You don't need both return and break. Your default should return a default theme.

hanskokx commented on lib/helpers/theme.dart:
```diff
@@ -0,0 +1,56 @@
+
+class CustomTheme {
+  static ThemeData mapTheme(dynamic theme) {
+    if (theme is int) {
```
I'll have to look into your enum, but passing an int seems odd.

hanskokx replied:
It's not very readable. If you came back to this in a year, how long would it take to figure it out?

tsonnen replied:
To me it was more clear to have the conversion from int to Enum in fewer places. 
`CustomTheme.mapTheme(service.get('ui_theme'))`
vs
`CustomTheme.mapTheme(ThemeEnum.values[service.get('ui_theme')])`
I don't feel strongly one way or the other. I was just trying to make the conversion from preference to theme more clear. I am open to suggestions though.

tsonnen replied:
I decided to change it so that the preference is stored as a string. Removes the need to map the preference to the enum. Should help to improve readability

hanskokx commented on lib/helpers/theme.dart:
```diff
@@ -0,0 +1,56 @@
+    return darkTheme; // personal preference
+  }
+
+  static var lightTheme = ThemeData(
```
var is lazy. Tell the good code what kind of variable this is.

tsonnen replied:
I was preferring var based on [DO follow a consistent rule for var and final on local variables.](https://dart.dev/guides/language/effective-dart/usage#do-follow-a-consistent-rule-for-var-and-final-on-local-variables) in the Dart style guide.

hanskokx replied:
I think you might have misconstrued the meaning. I'll have to take another look tomorrow though.

tsonnen replied:
Changing var to PrefServiceShared and running flutter analyze with pedantic checks flags it saying that types should be avoided for local variables.

I don't have a strong feeling one way or the other, if the style for this project is to include types for local variables that works for me and I can change it.

hanskokx replied:
I read over it again, and while I disagree with it, I will defer to it.  :)

hanskokx commented on lib/main.dart:
```diff
@@ -1,29 +1,36 @@
-void main() => runApp(MyApp());
+void main() async {
+  WidgetsFlutterBinding.ensureInitialized();
+  var service = await PrefServiceShared.init(
```
var should actually define the variable type we are expecting.

hanskokx commented on lib/screens/settings_screen.dart:
```diff
@@ -0,0 +1,39 @@
+      appBar: AppBar(
+        title: Text('Preferences'),
+      ),
+      body: PrefPage(
```
I think PreferencesPage would be a better name

tsonnen replied:
It looks like most of the other pages/screens are named "*Screen", I think that the names should be consistent. I can change it to PreferencesScreen though.

hanskokx replied:
I agree. Page is for the web. Screen is for apps. I'm still trying to get my own brain switched over. 😉

hanskokx commented on lib/screens/settings_screen.dart:
```diff
@@ -0,0 +1,39 @@
+          PrefRadio(
+            title: Text('Dark Theme'),
+            value: ThemeEnum.darkTheme.index,
+            pref: 'ui_theme',
```
🤔 Maybe this should be `theme: ` or something?

hanskokx commented on lib/screens/settings_screen.dart:
```diff
@@ -0,0 +1,39 @@
+                  .setTheme(CustomTheme.mapTheme(ThemeEnum.darkTheme));
+            },
+          ),
+          PrefRadio(
```
I don't like the name. Good naming is so important.

tsonnen replied:
This is the name defined in [Pref](https://pub.dev/packages/pref). Would be possible to create custom widgets if we want.

hanskokx replied:
Not necessary. I thought maybe it was a widget you created yourself. Carry on.

hanskokx commented on test/theme_test.dart:
```diff
@@ -0,0 +1,41 @@
+      expect(themeNotfier.getTheme(), CustomTheme.lightTheme);
+
+      themeNotfier.setTheme(CustomTheme.darkTheme);
+      expect(themeNotfier.getTheme(), CustomTheme.darkTheme);
```
You should create an element on the page and verify the colors actually change.

hanskokx reviewed (changes requested):
A solid start.

tsonnen commented:
I made most of the recommended changes, and added notes to the other ones.
I changed the preference to be saved as a string which avoids the enum mapping. I am open to suggestions on another way

One thing of note, it seems like Text Theming is based on localization options, so the test for if the theme changes is just checking that the brightness of the current theme matches what the brightness should be since just trying to compare the themes to each other failed. If there is a better way, I am up to change it, but this seemed like a good solution for the time being.

hanskokx commented on lib/helpers/theme.dart:
```diff
@@ -0,0 +1,48 @@
+        assert(false, 'Unknown theme $theme');
+        break;
+    }
+    return darkTheme; // personal preference
```
What about grabbing the light/dark theme of the device as the default?

tsonnen replied:
Oh, I like it. I think that I might change the item that we are changing to be the '[ThemeMode](https://api.flutter.dev/flutter/material/ThemeMode-class.html)', since that should allow for better/more readable code.

It looks like it is possible to [get the system theme](https://stackoverflow.com/a/56307575) itself, but I think that using the built-in stuff should work.
